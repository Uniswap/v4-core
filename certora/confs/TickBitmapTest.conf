{
	"files": [
		"certora/helpers/TickBitmapTest.sol"
	],
	"verify": "TickBitmapTest:certora/specs/LibrariesTest/TickBitmapTest.spec",
	"process": "emv",
	"smt_timeout": "1000",
	"prover_args": [
		"-mediumTimeout 50",
		"-depth 4",
		"-s [yices:def,cvc4:def]",
	],
	"precise_bitwise_ops": true,
	"server": "production",
	"solc_via_ir": true,
	"solc": "solc8.26",
	"solc_evm_version": "cancun",
	"msg": "UniswapV4 - TickBitmapTest",
}