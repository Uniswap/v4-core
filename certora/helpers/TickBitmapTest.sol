// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.20;

import { TickBitmap, BitMath } from "src/libraries/TickBitmap.sol";

contract TickBitmapTest {
    using TickBitmap for mapping(int16 => uint256);

    /// @dev The minimum tick that may be passed to #getSqrtPriceAtTick computed from log base 1.0001 of 2**-128
    int24 internal constant MIN_TICK = -887272;
    /// @dev The maximum tick that may be passed to #getSqrtPriceAtTick computed from log base 1.0001 of 2**128
    int24 internal constant MAX_TICK = 887272;

    /// @dev The minimum tick spacing value drawn from the range of type int16 that is greater than 0, i.e. min from the range [1, 32767]
    int24 internal constant MIN_TICK_SPACING = 1;
    /// @dev The maximum tick spacing value drawn from the range of type int16, i.e. max from the range [1, 32767]
    int24 internal constant MAX_TICK_SPACING = type(int16).max;

    int24 public tick0;
    int24 public tick1;
    int24 public tickSpacing0;
    mapping(int16 => uint256) public bitmap;

    int24 public nextTick; 
    bool public nextInitialized;
    int24 public compressed;
    int16 public wordPos;
    uint8 public bitPos;

    function nextToTick0Diff() public view returns (int256) {
        return int256(nextTick) - int256(tick0);
    }

    function isAtLeastOneWord(bool lte) public view returns (bool) {
        return (lte ? tick0 >= -8388352 : tick0 < 8388351);
    }

    function isValidTick() public view returns (bool) {
        return tick0 >= MIN_TICK && tick0 <= MAX_TICK;
    }

    function isValidTickSpacing() public view returns (bool) {
        return tickSpacing0 >= MIN_TICK_SPACING && tickSpacing0 <= MAX_TICK_SPACING;
    }

    function setTick0(uint24 absTick, bool isNegative) public {
        require (absTick <= uint24(MAX_TICK));
        tick0 = isNegative ? -int24(absTick) : int24(absTick);
    }

    function setTickSpacing0(uint24 tickSpacing) public {
        require (tickSpacing >= uint24(MIN_TICK_SPACING) && tickSpacing <= uint24(MAX_TICK_SPACING));
        tickSpacing0 = int24(tickSpacing);
    }

    function flipTick() external {
        bitmap.flipTick(tick0, tickSpacing0);
    }

    function flipTickSol() public {
        if (tick0 % tickSpacing0 != 0) revert TickBitmap.TickMisaligned(tick0, tickSpacing0);
        (int16 _wordPos, uint8 _bitPos) = TickBitmap.position(tick0 / tickSpacing0);
        uint256 mask = 1 << _bitPos;
        bitmap[_wordPos] ^= mask; 
    }

    function compress() public {
        compressed = TickBitmap.compress(tick0, tickSpacing0);
    }

    function compressSol() public {
        compressed = tick0 / tickSpacing0;
        if (tick0 < 0 && tick0 % tickSpacing0 != 0) compressed--;
    }

    function position() public {
        (wordPos, bitPos) = TickBitmap.position(compressed);
    }

    function positionSol() public {
        wordPos = int16(compressed >> 8);
        bitPos = uint8(uint24(compressed % 256));
    }

    function getBitPos(bool lte) public view returns (uint8) {
        (int16 wordPos_, uint8 bitPos_) = TickBitmap.position(
            TickBitmap.compress(tick0, tickSpacing0) + (lte ? int24(0) : int24(1))
        );
        return bitPos_;
    }

    function getBitPosNext() public view returns (uint8) {
        (int16 wordPos_, uint8 bitPos_) = TickBitmap.position(nextTick / tickSpacing0);
        return bitPos_;
    }

    function getBitMapAtWord(bool lte) public view returns (uint256) {
        (int16 wordPos_, ) = TickBitmap.position(
            TickBitmap.compress(tick0, tickSpacing0) + (lte ? int24(0) : int24(1))
        );
        return bitmap[wordPos_];
    }

    function getBitPosMask(bool lte, uint8 bitPos) public pure returns (uint256) {
        unchecked {
            return lte 
            ? type(uint256).max >> (uint256(type(uint8).max) - bitPos)
            : ~((1 << bitPos) - 1);
        }
    }

    function setNextInitializedTickWithinOneWord(bool lte) external {
        (nextTick, nextInitialized) = bitmap.nextInitializedTickWithinOneWord(tick0, tickSpacing0, lte);
    }

    function isInitialized0() public view returns (bool) {
        unchecked {
            if ((tick0/ tickSpacing0)*tickSpacing0 != tick0) return false;
            (int16 _wordPos, uint8 _bitPos) = TickBitmap.position(tick0 / tickSpacing0);
            return bitmap[_wordPos] & (1 << _bitPos) != 0;
        }
    }

    function isInitialized1() public view returns (bool) {
        unchecked {
            if ((tick1 / tickSpacing0)*tickSpacing0 != tick1) return false;
            (int16 _wordPos, uint8 _bitPos) = TickBitmap.position(tick1 / tickSpacing0);
            return bitmap[_wordPos] & (1 << _bitPos) != 0;
        }
    }

    function isNextInitialized() public view returns (bool) {
        unchecked {
            if ((nextTick / tickSpacing0)*tickSpacing0 != nextTick) return false;
            (int16 _wordPos, uint8 _bitPos) = TickBitmap.position(nextTick / tickSpacing0);
            return bitmap[_wordPos] & (1 << _bitPos) != 0;
        }
    }

    function tick1BetweenNextAndCurrent(bool lte) public view returns (bool) {
        return lte
            ? (tick1 > nextTick && tick1 <= tick0) 
            : (tick1 < nextTick && tick1 > tick0);
    }

    function nextTickGTTick() public view returns (bool) {
        return nextTick > tick0;
    }

    function differentTicks01() public view returns (bool) {
        return (tick0 != tick1 && tickSpacing0 > 0);
    }

    function getLSB(uint256 x) external pure returns (uint8) {
        return BitMath.leastSignificantBit(x);
    }
}
