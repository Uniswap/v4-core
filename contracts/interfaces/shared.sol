// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.10;

/// @notice Represents a change in the pool's balance of token0 and token1.
struct BalanceDelta {
    int256 amount0;
    int256 amount1;
}
