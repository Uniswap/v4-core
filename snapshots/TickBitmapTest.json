{
  "flipTick_flippingATickThatResultsInDeletingAWord": "5109",
  "flipTick_flippingFirstTickInWordToInitialized": "22209",
  "flipTick_flippingSecondTickInWordToInitialized": "5178",
  "nextInitializedTickWithinOneWord_lteFalse_forEntireWord": "2263",
  "nextInitializedTickWithinOneWord_lteFalse_justBelowBoundary": "2252",
  "nextInitializedTickWithinOneWord_lteFalse_onBoundary": "2281",
  "nextInitializedTickWithinOneWord_lteTrue_forEntireWord": "2258",
  "nextInitializedTickWithinOneWord_lteTrue_justBelowBoundary": "2242",
  "nextInitializedTickWithinOneWord_lteTrue_onBoundary_gas": "2287"
}