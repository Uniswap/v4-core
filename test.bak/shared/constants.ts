export const MIN_TICK = -887272
export const MAX_TICK = 887272
export const MAX_TICK_SPACING = 32767
export const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000'
