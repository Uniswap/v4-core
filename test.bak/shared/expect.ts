import { expect, use } from 'chai'
import { solidity } from 'ethereum-waffle'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'

use(solidity)
use(jestSnapshotPlugin())

export { expect }
