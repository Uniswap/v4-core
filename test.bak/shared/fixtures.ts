import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { TestERC20 } from '../../typechain/TestERC20'

interface TokensFixture {
  currency0: TestERC20
  currency1: TestERC20
  token2: TestERC20
}

export async function tokensFixture(): Promise<TokensFixture> {
  const tokenFactory = await ethers.getContractFactory('TestERC20')
  const tokenA = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenB = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20
  const tokenC = (await tokenFactory.deploy(BigNumber.from(2).pow(255))) as TestERC20

  const [currency0, currency1, token2] = [tokenA, tokenB, tokenC].sort((tokenA, tokenB) =>
    tokenA.address.toLowerCase() < tokenB.address.toLowerCase() ? -1 : 1
  )

  return { currency0, currency1, token2 }
}
