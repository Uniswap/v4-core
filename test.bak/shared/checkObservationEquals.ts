import { BigNumber, BigNumberish } from 'ethers'
import { expect } from './expect'

// helper function because we cannot do a simple deep equals with the
// observation result object returned from ethers because it extends array
export default function checkObservationEquals(
  {
    tickCumulative,
    blockTimestamp,
    initialized,
    secondsPerLiquidityCumulativeX128,
  }: {
    tickCumulative: BigNumber
    secondsPerLiquidityCumulativeX128: BigNumber
    initialized: boolean
    blockTimestamp: number
  },
  expected: {
    tickCumulative: BigNumberish
    secondsPerLiquidityCumulativeX128: BigNumberish
    initialized: boolean
    blockTimestamp: number
  }
) {
  expect(
    {
      initialized,
      blockTimestamp,
      tickCumulative: tickCumulative.toString(),
      secondsPerLiquidityCumulativeX128: secondsPerLiquidityCumulativeX128.toString(),
    },
    `observation is equivalent`
  ).to.deep.eq({
    ...expected,
    tickCumulative: expected.tickCumulative.toString(),
    secondsPerLiquidityCumulativeX128: expected.secondsPerLiquidityCumulativeX128.toString(),
  })
}
