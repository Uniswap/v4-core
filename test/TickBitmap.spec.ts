import { ethers } from 'hardhat'
import { TickBitmapTest } from '../typechain/TickBitmapTest'
import { expect } from './shared/expect'
import snapshotGasCost from '@uniswap/snapshot-gas-cost'

describe('TickBitmap', () => {
  let tickBitmap: TickBitmapTest

  beforeEach('deploy TickBitmapTest', async () => {
    const tickBitmapTestFactory = await ethers.getContractFactory('TickBitmapTest')
    tickBitmap = (await tickBitmapTestFactory.deploy()) as TickBitmapTest
  })

  async function initTicks(ticks: number[]): Promise<void> {
    for (const tick of ticks) {
      await tickBitmap.flipTick(tick)
    }
  }

  describe('#isInitialized', () => {
    it('is false at first', async () => {
      expect(await tickBitmap.isInitialized(1)).to.eq(false)
    })
    it('is flipped by #flipTick', async () => {
      await tickBitmap.flipTick(1)
      expect(await tickBitmap.isInitialized(1)).to.eq(true)
    })
    it('is flipped back by #flipTick', async () => {
      await tickBitmap.flipTick(1)
      await tickBitmap.flipTick(1)
      expect(await tickBitmap.isInitialized(1)).to.eq(false)
    })
    it('is not changed by another flip to a different tick', async () => {
      await tickBitmap.flipTick(2)
      expect(await tickBitmap.isInitialized(1)).to.eq(false)
    })
    it('is not changed by another flip to a different tick on another word', async () => {
      await tickBitmap.flipTick(1 + 256)
      expect(await tickBitmap.isInitialized(257)).to.eq(true)
      expect(await tickBitmap.isInitialized(1)).to.eq(false)
    })
  })

  describe('#flipTick', () => {
    it('flips only the specified tick', async () => {
      await tickBitmap.flipTick(-230)
      expect(await tickBitmap.isInitialized(-230)).to.eq(true)
      expect(await tickBitmap.isInitialized(-231)).to.eq(false)
      expect(await tickBitmap.isInitialized(-229)).to.eq(false)
      expect(await tickBitmap.isInitialized(-230 + 256)).to.eq(false)
      expect(await tickBitmap.isInitialized(-230 - 256)).to.eq(false)
      await tickBitmap.flipTick(-230)
      expect(await tickBitmap.isInitialized(-230)).to.eq(false)
      expect(await tickBitmap.isInitialized(-231)).to.eq(false)
      expect(await tickBitmap.isInitialized(-229)).to.eq(false)
      expect(await tickBitmap.isInitialized(-230 + 256)).to.eq(false)
      expect(await tickBitmap.isInitialized(-230 - 256)).to.eq(false)
    })

    it('reverts only itself', async () => {
      await tickBitmap.flipTick(-230)
      await tickBitmap.flipTick(-259)
      await tickBitmap.flipTick(-229)
      await tickBitmap.flipTick(500)
      await tickBitmap.flipTick(-259)
      await tickBitmap.flipTick(-229)
      await tickBitmap.flipTick(-259)

      expect(await tickBitmap.isInitialized(-259)).to.eq(true)
      expect(await tickBitmap.isInitialized(-229)).to.eq(false)
    })

    it('gas cost of flipping first tick in word to initialized', async () => {
      await snapshotGasCost(await tickBitmap.getGasCostOfFlipTick(1))
    })
    it('gas cost of flipping second tick in word to initialized', async () => {
      await tickBitmap.flipTick(0)
      await snapshotGasCost(await tickBitmap.getGasCostOfFlipTick(1))
    })
    it('gas cost of flipping a tick that results in deleting a word', async () => {
      await tickBitmap.flipTick(0)
      await snapshotGasCost(await tickBitmap.getGasCostOfFlipTick(0))
    })
  })

  describe('#nextInitializedTickWithinOneWord', () => {
    beforeEach('set up some ticks', async () => {
      // word boundaries are at multiples of 256
      await initTicks([-200, -55, -4, 70, 78, 84, 139, 240, 535])
    })

    describe('lte = false', async () => {
      it('returns tick to right if at initialized tick', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(78, false)
        expect(next).to.eq(84)
        expect(initialized).to.eq(true)
      })
      it('returns tick to right if at initialized tick', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(-55, false)
        expect(next).to.eq(-4)
        expect(initialized).to.eq(true)
      })

      it('returns the tick directly to the right', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(77, false)
        expect(next).to.eq(78)
        expect(initialized).to.eq(true)
      })
      it('returns the tick directly to the right', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(-56, false)
        expect(next).to.eq(-55)
        expect(initialized).to.eq(true)
      })

      it('returns the next words initialized tick if on the right boundary', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(255, false)
        expect(next).to.eq(511)
        expect(initialized).to.eq(false)
      })
      it('returns the next words initialized tick if on the right boundary', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(-257, false)
        expect(next).to.eq(-200)
        expect(initialized).to.eq(true)
      })

      it('returns the next initialized tick from the next word', async () => {
        await tickBitmap.flipTick(340)
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(328, false)
        expect(next).to.eq(340)
        expect(initialized).to.eq(true)
      })
      it('does not exceed boundary', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(508, false)
        expect(next).to.eq(511)
        expect(initialized).to.eq(false)
      })
      it('skips entire word', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(255, false)
        expect(next).to.eq(511)
        expect(initialized).to.eq(false)
      })
      it('skips half word', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(383, false)
        expect(next).to.eq(511)
        expect(initialized).to.eq(false)
      })

      it('gas cost on boundary', async () => {
        await snapshotGasCost(await tickBitmap.getGasCostOfNextInitializedTickWithinOneWord(255, false))
      })
      it('gas cost just below boundary', async () => {
        await snapshotGasCost(await tickBitmap.getGasCostOfNextInitializedTickWithinOneWord(254, false))
      })
      it('gas cost for entire word', async () => {
        await snapshotGasCost(await tickBitmap.getGasCostOfNextInitializedTickWithinOneWord(768, false))
      })
    })

    describe('lte = true', () => {
      it('returns same tick if initialized', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(78, true)

        expect(next).to.eq(78)
        expect(initialized).to.eq(true)
      })
      it('returns tick directly to the left of input tick if not initialized', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(79, true)

        expect(next).to.eq(78)
        expect(initialized).to.eq(true)
      })
      it('will not exceed the word boundary', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(258, true)

        expect(next).to.eq(256)
        expect(initialized).to.eq(false)
      })
      it('at the word boundary', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(256, true)

        expect(next).to.eq(256)
        expect(initialized).to.eq(false)
      })
      it('word boundary less 1 (next initialized tick in next word)', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(72, true)

        expect(next).to.eq(70)
        expect(initialized).to.eq(true)
      })
      it('word boundary', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(-257, true)

        expect(next).to.eq(-512)
        expect(initialized).to.eq(false)
      })
      it('entire empty word', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(1023, true)

        expect(next).to.eq(768)
        expect(initialized).to.eq(false)
      })
      it('halfway through empty word', async () => {
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(900, true)

        expect(next).to.eq(768)
        expect(initialized).to.eq(false)
      })
      it('boundary is initialized', async () => {
        await tickBitmap.flipTick(329)
        const { next, initialized } = await tickBitmap.nextInitializedTickWithinOneWord(456, true)

        expect(next).to.eq(329)
        expect(initialized).to.eq(true)
      })

      it('gas cost on boundary', async () => {
        await snapshotGasCost(await tickBitmap.getGasCostOfNextInitializedTickWithinOneWord(256, true))
      })
      it('gas cost just below boundary', async () => {
        await snapshotGasCost(await tickBitmap.getGasCostOfNextInitializedTickWithinOneWord(255, true))
      })
      it('gas cost for entire word', async () => {
        await snapshotGasCost(await tickBitmap.getGasCostOfNextInitializedTickWithinOneWord(1024, true))
      })
    })
  })
})
