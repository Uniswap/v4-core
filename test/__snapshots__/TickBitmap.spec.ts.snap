// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickBitmap #flipTick gas cost of flipping a tick that results in deleting a word 1`] = `22206`;

exports[`TickBitmap #flipTick gas cost of flipping first tick in word to initialized 1`] = `44118`;

exports[`TickBitmap #flipTick gas cost of flipping second tick in word to initialized 1`] = `27018`;

exports[`TickBitmap #nextInitializedTickWithinOneWord lte = false gas cost for entire word 1`] = `2572`;

exports[`TickBitmap #nextInitializedTickWithinOneWord lte = false gas cost just below boundary 1`] = `2572`;

exports[`TickBitmap #nextInitializedTickWithinOneWord lte = false gas cost on boundary 1`] = `2572`;

exports[`TickBitmap #nextInitializedTickWithinOneWord lte = true gas cost for entire word 1`] = `2570`;

exports[`TickBitmap #nextInitializedTickWithinOneWord lte = true gas cost just below boundary 1`] = `2880`;

exports[`TickBitmap #nextInitializedTickWithinOneWord lte = true gas cost on boundary 1`] = `2570`;
